pilot/pkg/networking/core/v1alpha3/tracing.go

```diff
@@ -0,0 +1,521 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha3
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	opb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	tracingcfg "github.com/envoyproxy/go-control-plane/envoy/config/trace/v3"
+	hpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	tracing "github.com/envoyproxy/go-control-plane/envoy/type/tracing/v3"
+	xdstype "github.com/envoyproxy/go-control-plane/envoy/type/v3"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	telemetrypb "istio.io/api/telemetry/v1alpha1"
+	"istio.io/istio/pilot/pkg/extensionproviders"
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/bootstrap/platform"
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/pkg/log"
+)
+
+// this is used for testing. it should not be changed in regular code.
+var clusterLookupFn = extensionproviders.LookupCluster
+
+func configureTracing(opts buildListenerOpts, hcm *hpb.HttpConnectionManager) {
+	spec := opts.push.Telemetry.EffectiveTelemetry(opts.proxy.ConfigNamespace, labels.Collection{opts.proxy.Metadata.Labels})
+	configureTracingFromSpec(spec, opts, hcm)
+}
+
+func configureTracingFromSpec(spec *telemetrypb.Telemetry, opts buildListenerOpts, hcm *hpb.HttpConnectionManager) {
+	meshCfg := opts.push.Mesh
+	proxyCfg := opts.proxy.Metadata.ProxyConfigOrDefault(opts.push.Mesh.DefaultConfig)
+
+	if len(spec.GetTracing()) == 0 {
+		if !meshCfg.EnableTracing {
+			log.Debugf("no valid tracing configuration found")
+			return
+		}
+		// use the prior configuration bits of sampling and custom tags
+		hcm.Tracing = &hpb.HttpConnectionManager_Tracing{}
+		configureSampling(hcm.Tracing, 0.0, proxyCfg)
+		configureCustomTags(hcm.Tracing, map[string]*telemetrypb.Tracing_CustomTag{}, proxyCfg)
+		if proxyCfg.GetTracing().GetMaxPathTagLength() != 0 {
+			hcm.Tracing.MaxPathTagLength = wrapperspb.UInt32(proxyCfg.GetTracing().MaxPathTagLength)
+		}
+		return
+	}
+
+	if len(spec.Tracing) > 1 {
+		log.Warnf("invalid number of tracing configurations provided; using first configuration found")
+	}
+
+	tracingCfg := spec.Tracing[0]
+
+	if tracingCfg.DisableSpanReporting {
+		return
+	}
+
+	// provider config
+	providerName := meshCfg.GetDefaultProviders().GetTracing()
+	if len(tracingCfg.Providers) > 0 {
+		// only one provider is currently supported, safe to take first
+		providerName = tracingCfg.Providers[0].Name
+	}
+
+	providerConfigured := false
+	for _, p := range meshCfg.ExtensionProviders {
+		if strings.EqualFold(p.Name, providerName) {
```

Comment:
the `EqualFold` is different from the logic used in the ext-authz that just use simple string equal. I'm not sure the what is the intended API but it should be consistent in the two places. It is even better if we can refactor a little bit to reuse the common code, but this can come later.

Also could you update the `processExtensionProvider` in `extauthz.go` to properly ignore the telemetry providers instead of returning errors.

Reply:
I've updated `processExtensionProvider` in about the most conservative way possible (as I'm not an expert on that code). We should find a more flexible solution there, but I'll let you tackle that.

`EqualFold` is mostly driven by my desire to not support case sensitivity in provider names. I want `Zipkin` to be the same provider as `zipkin`.